# ==============================================================================
# REGRESSOR COMMIT DIFF
# ==============================================================================
# File: accessible/mac/GeckoTextMarker.h
# Commit: 9bc12053de5e
# Full Hash: 9bc12053de5e6b071d41c81d82859d1567c80c42
# Author: Eitan Isaacson <eitan@monotonous.org>
# Date: 2020-07-03 03:56:55
# Regressor Bug: 1649575
# File Overlap Count: 2
# Description:
#   Bug 1649575 - Part 4: Introduce GeckoTextMarker and GeckoTextMarkerRange. r=morgan
#   
#   These classes do a few things:
#   1. They implement and abstract common text operations, for example:
#    (a) comparing two markers to know which one precedes the other.
# ==============================================================================

diff -r 0441d1553b6b -r 9bc12053de5e accessible/mac/GeckoTextMarker.h
--- /dev/null	Thu Jan 01 00:00:00 1970 +0000
+++ b/accessible/mac/GeckoTextMarker.h	Thu Jul 02 21:35:05 2020 +0000
@@ -0,0 +1,62 @@
+/* clang-format off */
+/* -*- Mode: Objective-C++; tab-width: 2; indent-tabs-mode: nil; c-basic-offset: 2 -*- */
+/* clang-format on */
+/* vim: set ts=2 et sw=2 tw=80: */
+/* This Source Code Form is subject to the terms of the Mozilla Public
+ * License, v. 2.0. If a copy of the MPL was not distributed with this
+ * file, You can obtain one at http://mozilla.org/MPL/2.0/. */
+
+#ifndef _GeckoTextMarker_H_
+#define _GeckoTextMarker_H_
+
+typedef CFTypeRef AXTextMarkerRef;
+typedef CFTypeRef AXTextMarkerRangeRef;
+
+namespace mozilla {
+namespace a11y {
+
+class AccessibleOrProxy;
+
+class GeckoTextMarker final {
+ public:
+  GeckoTextMarker(const AccessibleOrProxy& aContainer, int32_t aOffset)
+      : mContainer(aContainer), mOffset(aOffset) {}
+
+  GeckoTextMarker(const GeckoTextMarker& aPoint)
+      : mContainer(aPoint.mContainer), mOffset(aPoint.mOffset) {}
+
+  GeckoTextMarker(AccessibleOrProxy aDoc, AXTextMarkerRef aTextMarker);
+
+  GeckoTextMarker() : mContainer(nullptr), mOffset(0) {}
+
+  id CreateAXTextMarker();
+
+  bool operator<(const GeckoTextMarker& aPoint) const;
+
+  AccessibleOrProxy mContainer;
+  int32_t mOffset;
+};
+
+class GeckoTextMarkerRange final {
+ public:
+  GeckoTextMarkerRange(const GeckoTextMarker& aStart,
+                       const GeckoTextMarker& aEnd)
+      : mStart(aStart), mEnd(aEnd) {}
+
+  GeckoTextMarkerRange(AccessibleOrProxy aDoc,
+                       AXTextMarkerRangeRef aTextMarkerRange);
+
+  id CreateAXTextMarkerRange();
+
+  bool IsValid() const {
+    return !mStart.mContainer.IsNull() && !mEnd.mContainer.IsNull();
+  };
+
+  GeckoTextMarker mStart;
+  GeckoTextMarker mEnd;
+};
+
+}  // namespace a11y
+}  // namespace mozilla
+
+#endif